[red-knot] function parameter types (#14802)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/function/parameters.md

# Function parameter types

> Within a function scope, the declared type of each parameter is its annotated type (or Unknown if no
> annotation). The initial inferred type is the union of the declared type with the type of the

Reviewer: ```suggestion
Within a function scope, the declared type of each parameter is its annotated type (or Unknown if not
annotated). The initial inferred type is the union of the declared type with the type of the
```

default value expression (if any). If both are fully static types, this union should simplify to the
annotated type (since the default value type must be assignable to the annotated type, and for fully
static types this means subtype-of, which simplifies in unions). But if the annotated type is
Unknown or another non-fully-static type, the default value type may still be relevant as lower
bound.

The variadic parameter is a variadic tuple of its annotated type; the variadic-keywords parameter is
a dictionary from strings to its annotated type.

## Parameter kinds

```py
from typing import Literal

> def f(a, b: int, c=1, d: int = 2, /, e=3, f: Literal[4] = 4, *args: object, g=5, h: Literal[6] = 6, **kwargs: str) -> bytes:

Reviewer 2: nit: the return annotation is irrelevant here and is slightly distracting:

```suggestion
def f(a, b: int, c=1, d: int = 2, /, e=3, f: Literal[4] = 4, *args: object, g=5, h: Literal[6] = 6, **kwargs: str):
```

    reveal_type(a)  # revealed: Unknown
    reveal_type(b)  # revealed: int

>     reveal_type(c)  # revealed: Unknown | Literal[1]

Reviewer: Unrelated to this PR and I know you're aware of it.... I find it very confusing as a *user* that this infers to `Literal[1] | Unkown`, and I'd have a hard time explaining it to any colleague of why and even what it means.

Reviewer 3: I know this is not the answer you are looking for, but let me try to explain the current/intended behavior, to see if my own understanding is correct.

We want to follow a philosophy where a perfectly valid un-annotated program should ideally not lead us to issue any diagnostics.

Consider this perfectly reasonable and working (untyped) program:
```py
def f(x=0):
    return x * x + 1

f(1.2)
```
pyright issues a diagnostic here (*Argument of type "float" cannot be assigned to parameter "x" of type "int"*) because it infers a (too) narrow parameter type of `int`.

We infer `Unknown | Literal[0]` instead, which does not raise any diagnostic. We *also* wouldn't raise a diagnostic if you were to call `f("foo")`, but you can always achieve this by annotating your function parameter appropriately (e.g. `x: float`). You can opt into more safety.

On the other hand of the spectrum, consider this non-crashing — but obviously error-prone — program:
```py
def f(name, age=None):
     # lots of code
    print(f"Next year, {name} will turn {age+1}")

f("Max", 20)
```
mypy does not raise any diagnostic here, because it infers a (too) wide parameter type of `Any` for `age`.

We infer `Unknown | NoneType` instead. The meaning of this type is: an unknown set of Python objects, but *at least as large as* the set `{None}`. This allows us to issue a *Operator `+` not supported* diagnostic (not yet but eventually; only works for operators like `in` so far).

Author: I couldn't say it better than @sharkdp did, in terms of explaining what this type means and why we need it.

I understand the real issue here is that users who haven't gotten such a good explanation will likely be confused by these types, because other type checkers don't use them much. I'm not yet sure how to reconcile this. It may be that we can find a more intuitive notation, or it may be that this is just something we have to address via documentation.

Reviewer: Yeah. I just see myself struggling with it and am worried that people will accept that Red Knot shows these weird inferred types and maybe even understand why but without fully understanding what it means

    reveal_type(d)  # revealed: int
    reveal_type(e)  # revealed: Unknown | Literal[3]
    reveal_type(f)  # revealed: Literal[4]
    reveal_type(g)  # revealed: Unknown | Literal[5]
    reveal_type(h)  # revealed: Literal[6]

    # TODO: should be `tuple[object, ...]` (needs generics)
    reveal_type(args)  # revealed: tuple

    # TODO: should be `dict[str, str]` (needs generics)
    reveal_type(kwargs)  # revealed: dict
```

## Unannotated variadic parameters

...are inferred as tuple of Unknown or dict from string to Unknown.

```py
def g(*args, **kwargs):
    # TODO: should be `tuple[Unknown, ...]` (needs generics)
    reveal_type(args)  # revealed: tuple

    # TODO: should be `dict[str, Unknown]` (needs generics)
    reveal_type(kwargs)  # revealed: dict
```

## Annotation is present but not a fully static type

The default value type should be a lower bound on the inferred type.

```py
from typing import Any

def f(x: Any = 1):
    reveal_type(x)  # revealed: Any | Literal[1]

> ```

Reviewer: Can we add a test case that demonstrates the behavior if the annotated type and the default expression aren't compatible and how that impacts the uses?
